File: javascript/crypto-square/crypto-square.js
var Crypto = function(plainString) {
	this.rawText = plainString;
	this.normalizedText = this.normalizePlaintext();
};

// TODO: Once again, `replace()` probably hides a loop.
// Figure out if there's a better way to write this method.
Crypto.prototype.normalizePlaintext = function() {
	return this.rawText.toLowerCase().replace(/[-!#$%^&*()_+|~=`{}\[\]:";'<>?,.\/\s]*/g, '');
};

Crypto.prototype.size = function() {
	return Math.ceil(Math.sqrt(this.normalizedText.length));
};

Crypto.prototype._text_helper = function(textToSegment) {
	var segmentedText = [];
	
	var segmentSize = this.size();
	var loops = textToSegment.length / segmentSize;
	for (var i = 0; i < loops; ++i) {
		segmentedText.push(textToSegment.substr(i * segmentSize, segmentSize));
	}
	
	return segmentedText;
};

Crypto.prototype.plaintextSegments = function() {
	return this._text_helper(this.normalizedText);
};

Crypto.prototype.normalizeCiphertext = function() {
	var encryptedSegments = this._text_helper(this.ciphertext());
	
	// TODO: Is there a better way to do this? `join()` is very likely to hide a loop.
	return encryptedSegments.join(' ');
};

// FIXME: This method looks ugly. Needs refactoring.
Crypto.prototype.ciphertext = function() {
	var encryptedText = '';
	var plainSegments = this.plaintextSegments();
	
	var plainSegmentsLength = plainSegments.length;
	var segmentLength = plainSegments[0].length;
	for (var j = 0; j < segmentLength; ++j) {
		for (var i = 0; i < plainSegmentsLength; ++i) {
			if ('undefined' !== typeof plainSegments[i][j]) {
				encryptedText += plainSegments[i][j];
			}
		}
	}
	
	return encryptedText;
};

module.exports = Crypto;

File: python/crypto-square/crypto_square.py
import re
import math


def segment_text(clean_input, segment_size):
    segmented_text = []
    
    limit = len(clean_input) // segment_size
    print('limit = ', limit)
    for i in range(limit):
        segmented_text.append(clean_input[i * segment_size : segment_size ** (i + 1)])
    
    return segmented_text

def encode_helper(clean_input, segments):
    encoded_text = ''
    
    segments_len = len(segments)
    segment_len = len(segments[0])
    for j in range(segment_len):
        for i in range(segments_len):
            # if ('undefined' !== typeof plainSegments[i][j]) {
                encoded_text += segments[i][j]
            # }
        
        encoded_text += " "
    
    return encoded_text.rstrip()

def encode(message):
    if (0 == len(message)):
        return ''
    
    # TODO: clean up and normalize input
    clean_input = message.lower()
    regex = re.compile(r'[-!#$%^&*()_+|~=`{}\[\]:\";\'<>?,.\/\s]')
    clean_input = regex.sub('', clean_input)
    
    size = math.ceil(math.sqrt(len(clean_input)))
    segmented_text = segment_text(clean_input, size)

    encoded_text = encode_helper(clean_input, segmented_text)
    
    print('clean_input = ', clean_input)
    print('size = ', size)
    print('segmented_text = ', segmented_text)
    print('encoded_text = ', encoded_text)
    
    return encoded_text
    